### ISSUE TITLE
Cannot assign {|a: number|} to $ReadOnly<{a: number, b?: string}>

### ISSUE THREAD

**Issue by jedwards1211** [facebook/flow#6164]
There is no danger in this, but flow considers it an error:
[Try flow link](https://flow.org/try/#0MYewdgzgLgBAngLhgbwD4EMlgK4FsBGApgE6oC+MAvCpjAExkDcAUKJLAF5IAkASoegAmAeTAAbOAB5ktHARIAaGPgD8SaMQCWYAOZkAfFXgsgA)
```js
const y: {|a: number|} = {a: 2};
const z: $ReadOnly<{a: number, b?: string}> = y;
```
```
2: const z: $ReadOnly<{a: number, b?: string}> = y;
                                                 ^ Cannot assign `y` to `z` because property `b` is missing in object type [1] but exists in object type [2].
References:
1: const y: {|a: number|} = {a: 2};
            ^ [1]
2: const z: $ReadOnly<{a: number, b?: string}> = y;
                      ^ [2]
```

**Comment by lttb** [comment]
The problem here is that you're trying to cast the object with exact type without `b` property to the type that contains it (it's like when you want to cast the 'subType' to the 'Type'). As for me, it's an expected behavior.

But you can do things like this (conversely):

```js
const y: {|a: number, b?: string|} = {a: 2};
const z: $ReadOnly<{a: number}> = y; // no error
```

**Comment by jedwards1211** [comment]
@lttb you haven't given any explanation for *why* that would be a problem (would it allow me to do something that violates my type expectations?)

I can cast an object literal without `b` to one that has it as a maybe type:
```js
({a: 2}: {a: number, b?: number}) // no errors
```

My OP is safe because
* it's not possible for `z.b` to be something other than `string` (because using an exact type for `y` guarantees that it doesn't have a `b` fields)
* it's not possible to add a `b` field to `y` via `z.b = 'hello'` (because `z` is a `$ReadOnly` type)

Illustrations of how omitting these safeguards would be dangerous:

### `z` is not `$ReadOnly`
```js
const y: {|a: number|} = {a: 2}
const z: {a: number, b?: string} = y
// if the above assignment was allowed, then I would be able to set y.b:
z.b = 'hello'
```

### `y` is not an exact type
```js
const y: {a: number} = {a: 2, b: 3}
const z: $ReadOnly<{a: number, b?: string}> = y
if (z.b) console.log(z.b.toUpperCase()) // whoops, z.b is a number
```

**Comment by lttb** [comment]
I'm not sure the this is the right explanation, but this issue is about types compatibility, as I see. If you write something like e.g. `(var: Type)` or `(const x: Type = var)`, you also assert that the `Type` is compatible with the type of `var`.

With your description, we can make an assumption that any **exact** type with just **required** props should be compatible with any **read-only** type with *that required* props and any other **optional** (assuming that there are no conflicts or intersection in props names and types). I mean, following this logic, an expression like this should be valid, because we can't modify `z` and write in optional fields:

```js
type A = $Exact<{a, b, c, /*...*/}
type B = $ReadOnly<{a, b, c, /*...*/, a1?, b1?, c1?, /*... (and any other with optional props)*/>

declare var y: A
const z: B  = y
```

But if we'll take a look at the types, that means that that the type of `y` - `A` should be compatible with the type `B`, but it's not the truth at all. Yeah, we might see that it's safe in the case above for `y` and `z`, but not in general.

**Comment by lttb** [comment]
> I can perform the following assignment without problems, so why shouldn't I be able to do the cast in my OP?
> const y: $ReadOnly<{a: number, b?: string}> = {a: 2}

It works because flow infers the type of `{a: 2}` and casts it to the `$ReadOnly<{a: number, b?: string}>`, so there is no types compatibility problem. For example,
```js
const z = {a: 2} // infered to {|+a: number, +b?: string|}
const y: $ReadOnly<{a: number, b?: string}> = z
```

When you've got the strict types, flow should assert their compatibility too. You can make a subtype transformation this way: `{|a: number, b?: string|} -> {|+a: number|}`, but in the other direction will not work.

**Comment by jedwards1211** [comment]
I'm not following you, how is it not safe in general?

**Comment by jedwards1211** [comment]
In other words, they should make `$ReadOnly<T>` also a subtype of `$Exact<$ExcludeOptionalProps<T>>` (`$ExcludeOptionalProps` is not a real utility type, I made it up)

**Comment by lttb** [comment]
hi @jedwards1211! so it looks like it was fixed in v0.75 (see https://github.com/facebook/flow/blob/master/Changelog.md#0750), and `{| |}` acts as a subtype of `{ +p?: T }` now, so your example works well

**Comment by jedwards1211** [comment]
awesome, yup!  Thanks for letting me know :) I didn't notice
